### PR TITLE
openssl_pkcs12: do not crash when there's no certificate and/or private key in existing PKCS#12 file

### LINKED ISSUE #103: openssl_pkcs12: cannot re-create pkcs12 with no private key
Original issue from Ansible github: [https://github.com/ansible/ansible/issues/59945]()

##### SUMMARY

When generating a PKCS12 keystore without a private key, if the file exists, parsing it fails

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
openssl_pkcs12 (likely caused by <https://github.com/ansible-collections/community.crypto/blob/main/plugins/modules/openssl_pkcs12.py#L363>)

##### ANSIBLE VERSION

```paste below
ansible 2.9.6
  config file = /home/redacted/dev/redacted/ansible.cfg
  configured module search path = ['/home/redacted/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/redacted/.local/lib/python3.6/site-packages/ansible
  executable location = /home/redacted/.local/bin/ansible
  python version = 3.6.9 (default, Nov  7 2019, 10:44:02) [GCC 8.3.0]
```

##### CONFIGURATION

```paste below
ANSIBLE_NOCOWS(env: ANSIBLE_NOCOWS) = True
DEFAULT_CALLBACK_WHITELIST(/home/redacted/dev/redacted/ansible.cfg) = ['timer']
DEFAULT_FILTER_PLUGIN_PATH(/home/redacted/dev/redacted/ansible.cfg) = ['/home/redacted/dev/redacted/filter_plugins']
DEFAULT_LOG_PATH(/home/redacted/dev/redacted/ansible.cfg) = /var/log/ansible.log
DEFAULT_ROLES_PATH(/home/redacted/dev/redacted/ansible.cfg) = ['/home/redacted/dev/redacted/roles']
DEFAULT_STDOUT_CALLBACK(/home/redacted/dev/redacted/ansible.cfg) = yaml
```

##### OS / ENVIRONMENT

Ubuntu 18.04 inside WSL1 on Windows 10, but issue has been reproduced across various platforms (mainly CentOS 7.7)

##### STEPS TO REPRODUCE

```yaml

```

##### EXPECTED RESULTS

- Create a PFX/P12 file with only certificates ("other_certificates")
- Try to parse the file or run the export again over the same file

```yaml
openssl_pkcs12:
    path: "{{ truststore_path }}"
    name: truststore
    other_certificates: "{{ cacert_path }}"
```

##### ACTUAL RESULTS

PyOpenSSL fails to dump the private key (which is None when parsed from p12)

```paste below
      File "/tmp/ansible_openssl_pkcs12_payload_tfn6ujgx/ansible_openssl_pkcs12_payload.zip/ansible/modules/crypto/openssl_pkcs12.py", line 449, in <module>
      File "/tmp/ansible_openssl_pkcs12_payload_tfn6ujgx/ansible_openssl_pkcs12_payload.zip/ansible/modules/crypto/openssl_pkcs12.py", line 412, in main
      File "/tmp/ansible_openssl_pkcs12_payload_tfn6ujgx/ansible_openssl_pkcs12_payload.zip/ansible/modules/crypto/openssl_pkcs12.py", line 246, in check
      File "/tmp/ansible_openssl_pkcs12_payload_tfn6ujgx/ansible_openssl_pkcs12_payload.zip/ansible/modules/crypto/openssl_pkcs12.py", line 341, in parse
      File "/usr/lib/python3/dist-packages/OpenSSL/crypto.py", line 1899, in dump_privatekey
        raise TypeError("pkey must be a PKey")
    TypeError: pkey must be a PKey
```

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Fixes #103.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
openssl_pkcs12

### CODE REVIEW

File: plugins/modules/openssl_pkcs12.py [hunk 1]
```diff
@@ -360,10 +360,12 @@ def parse(self):
...
-            pkey = crypto.dump_privatekey(crypto.FILETYPE_PEM,
-                                          p12.get_privatekey())
-            crt = crypto.dump_certificate(crypto.FILETYPE_PEM,
-                                          p12.get_certificate())
+            pkey = p12.get_privatekey()
+            if pkey is not None:
+                pkey = crypto.dump_privatekey(crypto.FILETYPE_PEM, pkey)
+            crt = p12.get_certificate()
+            if crt is not None:
+                crt = crypto.dump_certificate(crypto.FILETYPE_PEM, crt)
```
Comment (abadger) [inline plugins/modules/openssl_pkcs12.py:368]: Does this mean that the caller also has to check whether the pkey and crt returned from parse() is None?

It looks like Line 448-450 might need to check whether they are None before formatting their output:
   https://github.com/ansible-collections/community.crypto/pull/109/files#diff-f695e3d82cb39aaa7fe1f0598978dd25R448

Comment (felixfontein) [reply plugins/modules/openssl_pkcs12.py:368]: Yep, that one should better be adjusted as well, so the "parsed" file doesn't start with `NoneNone` :) Fixed that and added a test.

**Comment by abadger** [comment]
As far as i can see without having actually run the code, this looks good now :-)

**Comment by felixfontein** [comment]
The new tests also pass (except the macOS failure in CI, but that is unrelated to this PR - I'll have to investigate what that is...), so it should be good :)

**Comment by felixfontein** [comment]
Looks like the failing CI was caused by cffi 1.14.3, which has been yanked. Unfortunately the pip version on macOS (8 something) does ignore that, since support for yanked packages was only added in pip 19.2...

**Comment by felixfontein** [comment]
(Rebased to remove test commits for fixing the macOS problem. That problem has now been fixed in `main`.)

**Comment by felixfontein** [comment]
@abadger thanks a lot for reviewing this!

**Pull request merged**